apps/music/js/metadata/album_art_cache.js

```diff
@@ -13,9 +13,67 @@ var AlbumArtCache = (function() {
   var THUMBNAIL_WIDTH = 300;
   var THUMBNAIL_HEIGHT = 300;
 
-  // The level one cache for thumbnails; maps a cache key (see below) to a blob:
-  // URL. There's also a level two cache, backed by asyncStorage.
-  var L1Cache = {};
+  // The level one caches for thumbnails and full-size images; maps a cache key
+  // (see below) to a blob: URL. There's also a level two cache for thumbnails,
+  // backed by asyncStorage.
+  var L1Cache = {'thumbnail': {}, 'fullsize': {}};
+
+  /**
+   * Get a URL for a full-sized version of the cover art for a given file (if
+   * any).
+   *
+   * @param {Object} fileinfo The info for the file we want album art for.
+   * @param {Boolean} noPlaceholder True if this function should only return an
+   *   image if there's cover art for the file. Otherwise, this function will
+   *   return a placeholder image if there's no cover art.
+   * @return {Promise} A promise returning either the URL for the cover art, or
+   *   null.
+   */
+  function getFullSizeURL(fileinfo, noPlaceholder) {
+    // If there's no picture info in the metadata, return null or a placeholder.
+    if (!fileinfo.metadata.picture) {
+      return Promise.resolve(
+        noPlaceholder ? null : getDefaultCoverURL(fileinfo)
+      );
+    }
+
+    // See if we've already made a URL for this album. If so, return it.
+    var cacheKey = makeCacheKey(fileinfo);
+    if (cacheKey && cacheKey in L1Cache.fullsize) {
+      return Promise.resolve(L1Cache.fullsize[cacheKey]);
+    }
+
+    // Otherwise, get the blob from the album art, create a URL for it, and
+    // return it in the Promise.
+    return getAlbumArtBlob(fileinfo).then(function(blob) {
+      return makeURL(cacheKey, blob, 'fullsize');
```

Comment:
Also consider defining FULLSIZE AND THUMBNAIL constants so you don't have to use string literals throughout your code. Saves the risk of typos.